File: JAVA/Validate Binary Search Tree.java
public class Solution {
    public boolean isValidBST(TreeNode root) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode n = root;
        int last = Integer.MIN_VALUE;
        while (n != null || !stack.empty()) {
            if (n != null) {
                stack.push(n);
                n = n.left;
            }
            else {
                n = stack.pop();
                // visit n
                if (n.val <= last) return false;
                last = n.val;
                n = n.right;
            }
        }
        return true;
    }
}

File: Validate Binary Search Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode *root) {
        stack<TreeNode*> stk;
        TreeNode *node = root;
        TreeNode *prev = NULL;
        while (!stk.empty() || node != NULL) {
            if (node != NULL) {
                stk.push(node);
                node = node->left;
            } else {
                node = stk.top();
                stk.pop();
                // visit
                if (prev != NULL && prev->val >= node->val)
                    return false;
                prev = node;
                node = node->right;
            }
        }
        return true;
    }
};
